Document getting started with Maven in more detail

First, forgive me for the really basic questions as it has been a million years since I've worked with Java projects and am a java noob 😅  I think some possibly basic maven setup information is missing in this guide, hopefully the information here can help figure that out, and I will happily update the getting started guides to be more total-java-noob friendly.

I tried following the getting started guides on the bolt docs (in particular, the [Socket Mode Getting Started Guide](https://slack.dev/java-slack-sdk/guides/getting-started-with-bolt-socket-mode)), but got stuck pretty quickly.

First in the "Project Setup" section:

> After you create your Maven project

There is nothing linked here, so I googled and followed the [Maven quickstart instructions](https://maven.apache.org/guides/getting-started/maven-in-five-minutes.html). I was able to create a maven project using `mvn generate` and then compile the basic project using `mvn package`. Finally, I could run the built project using `java -cp` - so I thought I was ok.

Next I added the dependencies listed under the Maven section of Project Setup of the Bolt Getting Started Guide. However, the last note in this section confused me:

> If you use Java-WebSocket library, add the artifact instead of tyrus-standalone-client.

I don't know if I'm using Java-WebSocket or not by default, so I don't know if I should add the relevant artifact and remove the tyrus-standalone-client 🤷 ❓ 

Next I moved on to the "Run Your Bolt App in 3 Minutes" section of the getting started guide. However, there is no Maven-specific instructions here - only instructions for changing a build.gradle file - so I skipped to copying the code for MyApp.java into my maven project.

Next section I moved to is "Start the App with Two Env Variables". I set the two environment variables, but the running instructions here are again gradle-specific and nothing Maven-specific. So I assumed the same package and running instructions relevant to the maven quickstart applied. However, running `mvn package` with the basic socket mode app code in place yields:

```
12:05:12 in ~/src/slapp-BoltJavaApp
➜ mvn package
[INFO] Scanning for projects...
[INFO]
[INFO] -----------------< ca.filmaj.boltjavaapp:boltjavaapp >------------------
[INFO] Building boltjavaapp 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-resources-plugin:3.0.2:resources (default-resources) @ boltjavaapp ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /Users/fmaj/src/slapp-BoltJavaApp/src/main/resources
[INFO]
[INFO] --- maven-compiler-plugin:3.8.0:compile (default-compile) @ boltjavaapp ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 1 source file to /Users/fmaj/src/slapp-BoltJavaApp/target/classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /Users/fmaj/src/slapp-BoltJavaApp/src/main/java/ca/filmaj/boltjavaapp/MyApp.java:[11,38] lambda expressions are not supported in -source 7
  (use -source 8 or higher to enable lambda expressions)
[INFO] 1 error
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.784 s
[INFO] Finished at: 2022-05-20T12:05:17-04:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project boltjavaapp: Compilation failure
[ERROR] /Users/fmaj/src/slapp-BoltJavaApp/src/main/java/ca/filmaj/boltjavaapp/MyApp.java:[11,38] lambda expressions are not supported in -source 7
[ERROR]   (use -source 8 or higher to enable lambda expressions)
[ERROR]
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

### The page URLs

* https://slack.dev/java-slack-sdk/guides/getting-started-with-bolt-socket-mode

### More Context

My mvn and java versions:

```
➜ mvn --version
Apache Maven 3.8.5 (3599d3414f046de2324203b78ddcf9b5e4388aa0)
Maven home: /usr/local/Cellar/maven/3.8.5/libexec
Java version: 11.0.15, vendor: Homebrew, runtime: /usr/local/Cellar/openjdk@11/11.0.15/libexec/openjdk.jdk/Contents/Home
Default locale: en_CA, platform encoding: UTF-8
OS name: "mac os x", version: "12.3.1", arch: "x86_64", family: "mac"
```

@filmaj Thanks for this feedback! 

First of all, my basic stance is that **this document is not for Java beginners but for Java developers that would like to know how to use this SDK**. For this reason, I believe that it's fine to skip some very details. That being said, I agree that it would be great if we make the documents even more friendly.

>However, the last note in this section confused me:
>If you use Java-WebSocket library, add the artifact instead of tyrus-standalone-client.
>I don't know if I'm using Java-WebSocket or not by default, so I don't know if I should add the relevant artifact and remove the tyrus-standalone-client 🤷 ❓

Both Tyrus and Java-WebSocket are optional dependencies of slack-api-package / bolt-socket-mode. The default recommendation is Tyrus as it is the reference implementation of the standard WebSocket client spec in Java. However, some developers may prefer using the alternative: Java-WebSocket. That's the case here. I agree that the part can be clearer.  Can you send a pull request for it?

>Next I moved on to the "Run Your Bolt App in 3 Minutes" section of the getting started guide. However, there is no Maven-specific instructions here - only instructions for changing a build.gradle file

Do you mean that we should add pom.xml part here? The document already mentioned how to configure a project with Maven in the previous "Project Setup" part. For this reason, I didn't repeat the same thing. If you think that this is not developer-friendly (frankly, I don't think so), we can add the dependencies part to the section as an improvement.

> - so I skipped to copying the code for MyApp.java into my maven project.

I don't think that majority of Maven developers won't be confused here but it's fine to make the part more friendly by adding pom.xml example or saying that "Of course, you can use Maven for setting this project up!".

>[ERROR] /Users/fmaj/src/slapp-BoltJavaApp/src/main/java/ca/filmaj/boltjavaapp/MyApp.java:[11,38] lambda expressions are not supported in -source 7 (use -source 8 or higher to enable lambda expressions)

Your build settings force the compilation with Java 1.7. It's a common practice to set source / target language version: https://github.com/slackapi/java-slack-sdk/blob/v1.22.1/pom.xml#L44-L45

This is not specific to this SDK at all but if you think that we should tell this to the document visitors to make the documents even more inclusive, it would be okay to add an extra sentence about this.

I do understand that some things might be confusing to you. If you are keen to make our documents even more friendly to everyone (beyond the common expectations to SDK docs), I am happy to work with you on updating the documents.

Working on a PR for this... @seratch I have another dumb question for you: how do I run the basic Bolt for Java socket mode app using maven? I can `mvn compile package` to compile into a .jar now after updating the compiler source/target to 1.8. But I am still a stuck 🙈 

I tried more explicitly configuring the maven jar plugin in my `pom.xml` to point to the entrypoint for the app:

```
		<groupId>org.apache.maven.plugins</groupId>
		<artifactId>maven-jar-plugin</artifactId>
		<version>3.1.0</version>
		<configuration>
			<archive>
				<manifest>
					<addClasspath>true</addClasspath>
					<mainClass>ca.filmaj.boltjavaapp.MyApp</mainClass>
				</manifest>
			</archive>
		</configuration>
```

and here's my app:

```
13:38:53 in ~/src/slapp-BoltJavaMavenApp
➜ cat src/main/java/ca/filmaj/boltjavaapp/MyApp.java
package ca.filmaj.boltjavaapp;

import com.slack.api.bolt.App;
import com.slack.api.bolt.socket_mode.SocketModeApp;

public class MyApp {
  public static void main(String[] args) throws Exception {
    // App expects an env variable: SLACK_BOT_TOKEN
    App app = new App();

    app.command("/hello", (req, ctx) -> {
      return ctx.ack(":wave: Hello!");
    });

    // SocketModeApp expects an env variable: SLACK_APP_TOKEN
    new SocketModeApp(app).start();
  }
}
```

But trying to run it via `java -jar` I still get errors around class path for bolt not being able to be found:

```
13:39:11 in ~/src/slapp-BoltJavaMavenApp
➜ java -jar target/boltjavaapp-1.0-SNAPSHOT.jar
Exception in thread "main" java.lang.NoClassDefFoundError: com/slack/api/bolt/App
	at ca.filmaj.boltjavaapp.MyApp.main(MyApp.java:9)
Caused by: java.lang.ClassNotFoundException: com.slack.api.bolt.App
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	... 1 more
```

I'm sure this is super basic stuff, sorry for the dumb question. I tried googling it for a bit but didn't get anywhere...

--- a/docs/guides/getting-started-with-bolt-socket-mode.md
+++ b/docs/guides/getting-started-with-bolt-socket-mode.md
@@ -30,7 +30,7 @@ Let's start building a Slack app using Bolt! This guide includes instructions on
 
 ### Maven
 
-After you create your Maven project, you need to add the **bolt** dependency to your `pom.xml` file. The **bolt** dependency is a framework-agnostic module. To enable [Socket Mode](https://api.slack.com/apis/connections/socket), the **bolt-socket-mode** library and its provided-scope dependencies are also required for your app.
+After you [create your Maven project](https://maven.apache.org/guides/getting-started/maven-in-five-minutes.html), you need to add the **bolt** dependency to your `pom.xml` file. The **bolt** dependency is a framework-agnostic module. To enable [Socket Mode](https://api.slack.com/apis/connections/socket), the **bolt-socket-mode** library and its provided-scope dependencies are also required for your app.
 
 ```xml
 <dependency>
@@ -53,9 +53,14 @@ After you create your Maven project, you need to add the **bolt** dependency to 
   <artifactId>tyrus-standalone-client</artifactId>
   <version>1.17</version>
 </dependency>
+<dependency>
+  <groupId>org.slf4j</groupId>
+  <artifactId>slf4j-simple</artifactId>
+  <version>{{ site.slf4jApiVersion }}</version>
+</dependency>
 ```
 
-If you use Java-WebSocket library, add the artifact instead of **tyrus-standalone-client**.
+By default, the `tyrus-standalone-client` dependency is used as the implementation to manage socket connections in the `bolt-socket-mode` artifact. If instead you would prefer to use the `Java-WebSocket` implementation, swap its artifact in instead of **tyrus-standalone-client**:
 
 ```xml
 <dependency>
@@ -63,6 +68,15 @@ If you use Java-WebSocket library, add the artifact instead of **tyrus-standalon
   <artifactId>Java-WebSocket</artifactId>
   <version>1.5.1</version>
 </dependency>
+```
+
+You will also need to ensure you set the compiler source and target to at least 1.8:
+
+```xml
+<properties>
+  <maven.compiler.source>1.8</maven.compiler.source>
+  <maven.compiler.target>1.8</maven.compiler.target>
+</properties>
 ```
 
 ### Gradle
@@ -74,6 +88,7 @@ dependencies {
   implementation("com.slack.api:bolt-socket-mode:{{ site.sdkLatestVersion }}")
   implementation("javax.websocket:javax.websocket-api:1.1")
   implementation("org.glassfish.tyrus.bundles:tyrus-standalone-client:1.17")
+  implementation("org.slf4j:slf4j-simple:{{ site.slf4jApiVersion }}")
 }
 ```
 
@@ -84,7 +99,7 @@ dependencies {
 
 **bolt-socket-mode** is a handy way to start your [Socket Mode](https://api.slack.com/apis/connections/socket) app. It allows developers to build a Slack app backend service by writing only a main method initializes **App** and establishes a WebSocket connection to the Socket Mode servers.
 
-#### build.gradle
+#### `build.gradle`
 
 The following build settings should be working as-is. Put it in the root directory of your project.
 
@@ -167,7 +182,12 @@ The default constructor expects the following two env variables exist when start
 
 If you prefer configuring an **App** in a different way, write some code to initialize **AppConfig** on your own.
 
-Anyway, set the two env variables and hit `gradle run` on your terminal. The command runs your main method. For more detailed logging, `gradle run -DslackLogLevel=debug` is also available.
+Set the two environment variables and run:
+
+- Gradle users: `gradle run` (for more detailed logging: `gradle run -DslackLogLevel=debug`),
+- Maven users: `mvn compile exec:java -Dexec.mainClass="hello.MyApp"` (for more detailed logging you can also provide the `-Dorg.slf4j.simpleLogger.defaultLogLevel=debug` flag)
+
+The command runs your main method.
 
 ```bash
 # Visit https://api.slack.com/apps to know these
@@ -175,10 +195,11 @@ export SLACK_BOT_TOKEN=xoxb-...your-own-valid-one
 export SLACK_APP_TOKEN=xapp-...your-own-valid-one
 
 # run the main function
+# gradle users should run:
 gradle run
+# maven users should run:
+mvn compile exec:java -Dexec.mainClass="hello.MyApp"
 ```
-
-You will see the message saying "**⚡️ Bolt app is running!**" in stdout.
 
 If you get stuck this setup, go through the following checklist:
 
@@ -276,10 +297,6 @@ export SLACK_APP_TOKEN=xapp-...your-own-valid-one
 # run the main function
 gradle run
 ```
-
-... Did you see the message saying "**⚡️ Bolt app is running!**" in stdout?
-
-If yes, that's all settled! 🎉
 
 From here, all you need to do is write code and restart the app. Enjoy Bolt app development in Kotlin! 👋
 

--- a/docs/guides/getting-started-with-bolt.md
+++ b/docs/guides/getting-started-with-bolt.md
@@ -31,7 +31,7 @@ Let's start building a Slack app using Bolt! This guide includes instructions on
 
 ### Maven
 
-After you create your Maven project, you need to add the **bolt** dependency to your `pom.xml` file. The **bolt** dependency is a framework-agnostic module. If you use Bolt along with [Spring Boot](https://spring.io/projects/spring-boot), [Quarkus (Undertow)](https://quarkus.io/), and any others on top of Servlet environment, the **bolt-servlet** library is required for your app. Adding only **bolt-servlet** also works for you.
+After you [create your Maven project](https://maven.apache.org/guides/getting-started/maven-in-five-minutes.html), you need to add the **bolt** dependency to your `pom.xml` file. The **bolt** dependency is a framework-agnostic module. If you use Bolt along with [Spring Boot](https://spring.io/projects/spring-boot), [Quarkus (Undertow)](https://quarkus.io/), or any others on top of Servlet environment, the **bolt-servlet** library is required for your app. Adding only **bolt-servlet** also works for you.
 
 ```xml
 <dependency>
@@ -43,6 +43,11 @@ After you create your Maven project, you need to add the **bolt** dependency to 
   <groupId>com.slack.api</groupId>
   <artifactId>bolt-servlet</artifactId>
   <version>{{ site.sdkLatestVersion }}</version>
+</dependency>
+<dependency>
+  <groupId>org.slf4j</groupId>
+  <artifactId>slf4j-simple</artifactId>
+  <version>{{ site.slf4jApiVersion }}</version>
 </dependency>
 ```
 
@@ -56,6 +61,15 @@ If you run the Bolt app on the Jetty HTTP server without any frameworks, you can
 </dependency>
 ```
 
+You will also need to ensure you set the compiler source and target to at least 1.8:
+
+```xml
+<properties>
+  <maven.compiler.source>1.8</maven.compiler.source>
+  <maven.compiler.target>1.8</maven.compiler.target>
+</properties>
+```
+
 ### Gradle
 
 After you [create your Gradle project](https://docs.gradle.org/current/samples/sample_building_java_applications.html), add the **bolt** dependencies to `build.gradle`.
@@ -65,6 +79,7 @@ dependencies {
   implementation("com.slack.api:bolt:{{ site.sdkLatestVersion }}")
   implementation("com.slack.api:bolt-servlet:{{ site.sdkLatestVersion }}")
   implementation("com.slack.api:bolt-jetty:{{ site.sdkLatestVersion }}")
+  implementation("org.slf4j:slf4j-simple:{{ site.slf4jApiVersion }}")
 }
 ```
 
@@ -73,9 +88,9 @@ dependencies {
 
 ### Use **bolt-jetty**
 
-**bolt-jetty** is a handy way to start your Slack app server. It allows developers to build a Slack app backend service by writing only a main method initializes **App** and starts an HTTP server.
+**bolt-jetty** is a handy way to start your Slack app server. It allows developers to build a Slack app backend service by writing only a main method that initializes **App** and starts an HTTP server.
 
-#### build.gradle
+#### `build.gradle`
 
 The following build settings should be working as-is. Put it in the root directory of your project.
 
@@ -158,7 +173,12 @@ The default constructor expects the following two env variables exist when start
 
 If you prefer configuring an **App** in a different way, write some code to initialize **AppConfig** on your own.
 
-Anyway, set the two env variables and hit `gradle run` on your terminal. The command runs your main method. For more detailed logging, `gradle run -DslackLogLevel=debug` is also available.
+Set the two environment variables and run:
+
+- Gradle users: `gradle run` (for more detailed logging: `gradle run -DslackLogLevel=debug`),
+- Maven users: `mvn compile exec:java -Dexec.mainClass="hello.MyApp"` (for more detailed logging you can also provide the `-Dorg.slf4j.simpleLogger.defaultLogLevel=debug` flag)
+
+The command runs your main method.
 
 ```bash
 # Visit https://api.slack.com/apps to know these
@@ -166,7 +186,10 @@ export SLACK_BOT_TOKEN=xoxb-...your-own-valid-one
 export SLACK_SIGNING_SECRET=123abc...your-own-valid-one
 
 # run the main function
+# gradle users should run:
 gradle run
+# maven users should run:
+mvn compile exec:java -Dexec.mainClass="hello.MyApp"
 ```
 
 You will see the message saying "**⚡️ Bolt app is running!**" in stdout.

